spirv-fuzz: Consider all ids from dead blocks irrelevant

stefanomil opened the pull request:
This PR modifies the FactManager methods IdIsIrrelevant and GetIrrelevantIds so
that an id is always considered irrelevant if it comes from a dead block.

Fixes #3733 .

afd commented:
@stefanomil Looks like you have an unused parameter - likely one that's used only for assertion checking.

afd commented on source/fuzz/fact_manager/dead_block_facts.h:
```diff
@@ -33,6 +33,9 @@ class DeadBlockFacts {
   // See method in FactManager which delegates to this method.
   bool BlockIsDead(uint32_t block_id) const;
 
+  // Returns an unordered set of all the block ids that have been declared dead.
```
Nit: I try to avoid comments that reiterate implementation details of the return type.  I'd say here "Returns the set of all block ids that have been declared dead."  The fact that it's an unordered set is apparent from the type.

stefanomil replied:
Fixed

afd commented on source/fuzz/fact_manager/irrelevant_value_facts.cpp:
```diff
@@ -60,13 +62,39 @@ bool IrrelevantValueFacts::PointeeValueIsIrrelevant(uint32_t pointer_id) const {
+  // Get all the ids declared in dead blocks.
+  for (uint32_t block_id : dead_block_facts.GetDeadBlocks()) {
+    auto block = fuzzerutil::MaybeFindBlock(context, block_id);
+    if (!block) {
```
I think it would be good to remark here that it's possible for the block to no longer exist in the module, e.g. if it has been merged with another block.

stefanomil replied:
Done

afd reviewed (changes requested):
Thanks for making this big change!

A couple of nits in the comments.

Also: while it's great you modified all the existing tests to reflect the changes in the fact manager interface, I don't think you added any new tests to check whether ids from dead blocks are actually regarded as irrelevant.

Would you be able to do something for this?  There should already be tests that involve the BlockIsDead fact; you could adapt them to also assert that ids declared in those blocks are deemed irrelevant.

Once this is done I will approve and merge.

Vasniktel commented on source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp:
```diff
@@ -53,11 +53,15 @@ bool DataSynonymAndIdEquationFacts::OperationEquals::operator()(
 
 void DataSynonymAndIdEquationFacts::AddFact(
     const protobufs::FactDataSynonym& fact,
+    const DeadBlockFacts& dead_block_facts,
     const IrrelevantValueFacts& irrelevant_value_facts,
```
This looks a bit convoluted to me: we are not using `dead_block_facts` anywhere in the method and only passing it down into `irrelevant_value_facts.IdIsIrrelevant` call. Maybe it's better to refactor all submanagers into class fields instead? @afd, @stefanomil, what do you think?

stefanomil replied:
Summary of the discussion: we will keep it as it is (at least for now).

stefanomil commented:
I have made the changes and added tests to check that ids from dead blocks are considered irrelevant.

I have also changed it so that if a pointer is in a dead block, the pointee value is considered irrelevant rather than the pointer id. This is to be consistent with the checks made when declaring new irrelevant ids or pointee values.

Vasniktel commented on source/fuzz/fact_manager/irrelevant_value_facts.cpp:
```diff
@@ -56,17 +58,79 @@ void IrrelevantValueFacts::AddFact(
 
-bool IrrelevantValueFacts::PointeeValueIsIrrelevant(uint32_t pointer_id) const {
-  return pointers_to_irrelevant_pointees_ids_.count(pointer_id) != 0;
+bool IrrelevantValueFacts::PointeeValueIsIrrelevant(
```
I'm not sure if we can consider pointees irrelevant even if the corresponding instructions are in dead blocks. Consider:
```
%1 = OpFunction %void None %type
%2 = OpLabel
%3 = OpVariable ...
...
     OpBranchConditional %true %4 %dead
%4 = OpLabel
...
%dead = OpLabel
%5 = OpCopyObject %pointer %3 ; can't be marked with PointeeValueIsIrrelevant
...
```

This is related to the fact that we can't determine whether two pointers point to the same memory location right now (#3723).

stefanomil replied:
You are quite right, I have now fixed it so that they are not considered irrelevant

Vasniktel reviewed (changes requested):
Spotted a potential bug regarding irrelevant pointees.

afd reviewed (approved):
LGTM.
